### PR TITLE
SOLR-15274: QueryElevationComponent, no data dir

### PR DESCRIPTION
**Pull request by dsmiley**
https://issues.apache.org/jira/browse/SOLR-15274

* Remove support for a config in a data dir. Obsolete.
* Remove support for "versioned" (suffix) config
* Add auto config reloading after a commit (previously was only for the data dir).  SolrCloud not supported.

TODO Solr Ref Guide

### CODE REVIEW

**Comment by dsmiley** [comment]
The getConfigVersion method ought to move to something similar on SolrResourceLoader.  SRL is one of Solr's major APIs (albeit internal) so changing it deserves a separate issue, which I'll do.

File: solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java [hunk 1]
```diff
@@ -290,70 +275,83 @@ protected void handleInitializationException(Exception exception, Initialization
...
-    synchronized (elevationProviderCache) {
-      ElevationProvider elevationProvider;
-      elevationProvider = elevationProviderCache.get(null);
-      if (elevationProvider != null) return elevationProvider;
-
-      elevationProvider = elevationProviderCache.get(reader);
-      if (elevationProvider == null) {
-        Exception loadingException = null;
-        boolean resourceAccessIssue = false;
+  ElevationProvider getElevationProvider(IndexReader reader, SolrCore core) {
```
Comment (bruno-roustant) [inline solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java:311]: Nice refactoring!

**Review by bruno-roustant — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java [hunk 2]
```diff
@@ -218,61 +233,31 @@ private void parseUseConfiguredOrderForElevations() {
...
+      RefCounted<SolrIndexSearcher> searchHolder = null;
+      try {
+        searchHolder = core.getNewestSearcher(false);
+        if (searchHolder != null) {
+          IndexReader reader = searchHolder.get().getIndexReader();
+          getElevationProvider(reader, core); // computes and caches or throws
+          return cacheElevationProvider.size();
         }
+      } finally {
+        if (searchHolder != null) searchHolder.decref();
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java:256]: I didn't write this; merely moved around in the PR.  I think this one-liner is kind of common for searchHolder related stuff.  Even Google-Java-Format is fine with it.

File: solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java [hunk 3]
```diff
@@ -290,70 +275,83 @@ protected void handleInitializationException(Exception exception, Initialization
...
-   * @throws E If the exception is not absorbed.
+   * @param e The exception caught.  It will extend {@link IOException} if there was a resource
+   *          access issue.
+   * @return The {@link ElevationProvider} to use if the exception is absorbed (vs re-thrown).
    */
-  protected <E extends Exception> ElevationProvider handleConfigLoadingException(E e, boolean resourceAccessIssue) throws E {
-    throw e;
+  protected <E extends Exception> ElevationProvider handleConfigLoadingException(E e) {
+    if (cacheElevationProvider != null) { // thus at runtime (a search is in-progress)
+      String msg = e.toString(); // declare to avoid log isEnabled check
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java:288]: See `gradle/validation/validate-log-calls.gradle` Erick Erickson insisted we guard many of our log calls with a log enabled check. LUCENE-7788
IMO there shouldn't be checks on an error level but Erick had stronger opinions than I did at the time :-)

File: solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java [hunk 4]
```diff
@@ -143,11 +144,25 @@
 
   protected boolean initialized;
 
+  private final Object LOCK = new Object(); // for cache*
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java:147]: It's final & immutable so all-caps seems reasonable to me

**Pull request merged**